File: whatsnext/whatsnext.js
whatsnext = {
  schedule:function(){return schedule2018_19()},
  infoFunc:function(period){
    return {}
  },
  setInfoFunction:function(callback){
    if(typeof callback == "function"){
      this.infoFunc = callback
    }
  },
  scheduleToday:null,
  scheduleTodayDate:null,
  getScheduleToday: function(now){
    if(!now) now = new Date()
    if(this.scheduleToday != null){
      if(this.scheduleTodayDate == now.toDateString()){
        return this.scheduleToday
      }
    }

    var schedule = schedule2018_19()
    var days = ["sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"]
    var day = days[now.getDay()]
    for(let {name, date} of schedule.minimum_days){
      if(date.toDateString() == now.toDateString()){
        day =  "minimum"//[{name:name, dates:dates}, "now"]
      }
    }
    console.log("Today is a "+day)
    var todaysObject = JSON.parse(JSON.stringify(schedule[day]))
    //console.log(todaysObject)
    this.scheduleToday = {}

    function timeToday(object, now){
      let otherNow = new Date(now.toString())
      //console.log("Other Now for timeToday is "+otherNow)
      otherNow.setHours(object.hour, object.minute, 0)
      return otherNow
    }

    for(period in todaysObject){
      let periodObject = todaysObject[period]
      //console.log(periodObject)
      periodObject.start = timeToday(periodObject.start, now)
      periodObject.end = timeToday(periodObject.end, now)
      if(!isNaN(parseInt(period))){
        periodObject.info = this.infoFunc(period)
      }else{
        periodObject.info = null
      }
      //periodObject.info = periodObject.info
    }
    //console.log("Today's shcedule is: \n", todaysObject)
    this.scheduleToday = todaysObject
    this.scheduleTodayDate = now.toDateString()
    return this.scheduleToday
  },
  getThisPeriod:function(now){
    if(!now) now = new Date()
    var scheduleToday = this.getScheduleToday(now)
    for(period in scheduleToday){
      let periodObj = scheduleToday[period]
      if(periodObj.end > now){
        if(periodObj.start <= now){
          return periodObj
        }
      }
    }
    return {period:"none", start:null, end:null, info:null}
  },
  getNextPeriod:function(now){
    if(!now) now = new Date()
    var endOfThisPeriod = this.getThisPeriod(now).end
    if(!endOfThisPeriod) return {period:"none", start:null, end:null, info:null}
    var fourMinutesFromEndOfThisPeriod = new Date(endOfThisPeriod.valueOf()+(4*60*1000))
    var thisPeriod = this.getThisPeriod(fourMinutesFromEndOfThisPeriod)
    return thisPeriod
  },
  isSchool: function(now){
    if(!now) now = new Date()
    var schedule = this.schedule()

    // Is it summer break
    if(now < schedule.school_year.start || now > schedule.school_year.end){
      return [false, "Summer!!"]
    }

    // is it weekend
    var days = ["sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"]
    var day = days[now.getDay()]
    if(day == "saturday" || day == "sunday"){
      return [false, "Weekend"];
    }

    // is it a holiday or recess
    var nextDayOff = this.getNextDayOff(now)
    if(nextDayOff.rel == "now"){
      return [false, nextDayOff.name];
    }

    // is it before or after school
    var scheduleToday = this.getScheduleToday(now)
    let first = (scheduleToday[1] || scheduleToday[2])
    let last = (scheduleToday[8] || scheduleToday[7])
    var start = first.start
    var end = last.end
    if(now < start) return [false, "Before School"];
    if(now > end) return [false, "After School"];

    return [true, "Yup"];
  },
  getNextDayOff:function(now){
    if(!now) now = new Date();
    var schedule = this.schedule()
    for(let {name, date} of schedule.days_off){
      if(date > now){
        return {name:name, date:date, rel:"next"} 
      }
      if(date.toDateString() == now.toDateString()){
          return {name:name, date:date, rel:"now"}
      }
    }
    return "nope"
  },
  countdown:function(now, then){
    var diff = then.getTime() - now.getTime();

    var days = Math.floor(diff / (1000 * 60 * 60 * 24));
    diff -=  days * (1000 * 60 * 60 * 24);

    var hours = Math.floor(diff / (1000 * 60 * 60));
    diff -= hours * (1000 * 60 * 60);

    var mins = Math.floor(diff / (1000 * 60));
    diff -= mins * (1000 * 60);

    var seconds = Math.floor(diff / (1000));
    diff -= seconds * (1000);

    var final = ""
    if(days) final+=days+" days, "
    if(hours) final+=hours+" hours, "
    if(mins) final+=mins+" minutes, and "
    final+=seconds +" seconds"
    return final
  }
}
//whatsnext.getScheduleToday(new Date())

File: whatsnext/whatsnext.py
import datetime
import json

class whatsnext():
    def __init__(self, datetimeIn = (None)):
        self.time = datetimeIn
        if(type(datetimeIn) != datetime.datetime):
            self.time = datetime.datetime.today()
        self.schedule_base = (json.loads(open("schedule2018_19.json").read()))
        self.periodInfo = {}
        self.scheduleToday = None
        self.scheduleToday = self.schedule()

    def now(self):
        if(self.time):
            return self.time
        else:
            return datetime.datetime.today()

    def dictToDate(self, dict):
        if "year" in dict:
            return datetime.date(dict["year"], dict["month"], dict["day"])
        elif "minute" in dict:
            return datetime.time(dict["hour"], dict["minute"])

    def schedule(self):
        now = self.now()
        schedule = self.schedule_base
        if self.time and now.date() == self.time.date() and self.scheduleToday:
            return self.scheduleToday
        self.scheduleToday = False
        state = {"day":"monday"} # TODO: self.state
        day = state["day"]
        today_base = schedule[day]
        todaysObject = {}
        periodObj = None
        for period in today_base:
            periodObj_base = today_base[period]
            todaysObject[period] = {}
            periodObj = todaysObject[period]
            periodObj["start"] = self.dictToDate(periodObj_base["start"])
            periodObj["end"] = self.dictToDate(periodObj_base["end"])
            if periodObj_base["info"] != None:
                periodObj["info"] = {}
                if period in self.periodInfo:
                    periodObj["info"] = self.periodInfo[period]
                periodObj["info"]["period"] = period
        if todaysObject == {}:
            return None
        self.scheduleToday = todaysObject

        return self.scheduleToday

    @property
    def state(self):
        now = self.now()
        schedule = self.schedule_base
        state = {
            "day":None,
            "nextDayOff":None,
            "thisPeriod":None,
            "nextPeriod":None
        }

        if now.date() < self.dictToDate(schedule["school_year"]["start"]) or now.date() > self.dictToDate(schedule["school_year"]["end"]):
            state["day"] = "summer"

        days = ["monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"]
        day = days[now.weekday()]
        if day == "saturday" or day == "sunday":
            if state["day"] == None:
                state["day"] = "weekend"
        if state["day"] == None:
            state["day"] = day

        nextDayOff = None
        for obj in schedule["days_off"]:
            name = obj["name"]
            date = obj["date"]
            date = self.dictToDate(date)
            if(date > now.date()):
                nextDayOff = {"name":name, "date":date, "rel":"next"}
                break
            if(date == now.date()):
                nextDayOff = {"name":name, "date":date, "rel":"now"}
                break
        state["nextDayOff"] = nextDayOff

        if self.scheduleToday:
            scheduleToday = self.schedule()
            first = None
            last = None
            if "2" in scheduleToday:
                first = scheduleToday["2"]
            if "1" in scheduleToday:
                first = scheduleToday["1"]
            if "7" in scheduleToday:
                last = scheduleToday["7"]
            if "8" in scheduleToday:
                last = scheduleToday["8"]

            start = first["start"]
            end = last["end"]
            if now.time() < start:
                state["thisPeriod"] = "before school"
                state["nextPeriod"] = "before school"
            if now.time() > end:
                state["thisPeriod"] = "after school"
                state["nextPeriod"] = "after school"

            thisPeriod = state["thisPeriod"]
            for period in scheduleToday:
                periodObj = scheduleToday[period]
                if periodObj["end"] > now.time():
                    if periodObj["start"] <= now.time():
                        thisPeriod = periodObj
            state["thisPeriod"] = thisPeriod

            if(type(thisPeriod) == type({"imADictionary":True})):
                nextPeriod = state["nextPeriod"]
                fourMinutesFromEndOfThisPeriod = (now+datetime.timedelta(minutes=4)).time()
                for period in scheduleToday:
                    periodObj = scheduleToday[period]
                    if periodObj["end"] > fourMinutesFromEndOfThisPeriod:
                        if periodObj["start"] <= fourMinutesFromEndOfThisPeriod:
                            thisPeriod = periodObj
                state["nextPeriod"] = thisPeriod

        return state


if __name__ == "__main__":
    inst = whatsnext(datetime.datetime(2018, 11, 23, 10))
    print(inst.state)
